src/main/java/org/folio/fqm/service/EntityTypeFlatteningService.java

```diff
@@ -0,0 +1,242 @@
+import org.folio.fqm.exception.EntityTypeNotFoundException;
+import org.folio.fqm.repository.EntityTypeRepository;
+import org.folio.querytool.domain.dto.EntityType;
+import org.folio.querytool.domain.dto.EntityTypeColumn;
+import org.folio.querytool.domain.dto.EntityTypeSource;
+import org.folio.querytool.domain.dto.EntityTypeSourceJoin;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+
+@Service
+@RequiredArgsConstructor
+@Log4j2
+public class EntityTypeFlatteningService {
+  private final EntityTypeRepository entityTypeRepository;
+  private final ObjectMapper objectMapper;
+
+  // TODO: clean up
+  public Optional<EntityType> getFlattenedEntityType(UUID entityTypeId, boolean doFinalRenames) {
+    EntityType originalEntityType = entityTypeRepository
+      .getEntityTypeDefinition(entityTypeId)
+      .orElseThrow(() -> new EntityTypeNotFoundException(entityTypeId));
+    EntityType flattenedEntityType = new EntityType()
+      .id(originalEntityType.getId())
+      .name(originalEntityType.getName())
+      ._private(originalEntityType.getPrivate())
+      .defaultSort(originalEntityType.getDefaultSort())
+      .columns(originalEntityType.getColumns()) // TODO: probably remove
+      .idView(originalEntityType.getIdView())
+      .customFieldEntityTypeId(originalEntityType.getCustomFieldEntityTypeId())
+      .labelAlias(originalEntityType.getLabelAlias())
+      .root(originalEntityType.getRoot())
+      .sourceView(originalEntityType.getSourceView()) // Possibly unneeded
+      .sourceViewExtractor(originalEntityType.getSourceViewExtractor()); // Possibly unneeded
+
+    List<EntityTypeColumn> finalColumns = new ArrayList<>();
+    for (EntityTypeSource source : originalEntityType.getSources()) {
+      if (source.getType().equals("db")) {
+        Pair<EntityTypeSource, List<EntityTypeColumn>> updatePair = getConvertedSourceAndColumns(originalEntityType, source, null, false); // TODO: think about this, may not be able to hardcode false here
+        flattenedEntityType.addSourcesItem(updatePair.component1());
+        finalColumns.addAll(updatePair.component2());
+      } else {
+        UUID sourceEntityTypeId = UUID.fromString(source.getId());
+        EntityType flattenedSourceDefinition = getFlattenedEntityType(sourceEntityTypeId, false)
+          .orElseThrow(() -> new EntityTypeNotFoundException(sourceEntityTypeId));
+
+        // If an entity type source contains multiple db sources, then we need to keep the original alias in order to
+        // distinguish the different targets. Frequently, it will likely only have one db source. In this case we
+        // can use the outer alias only, in order to keep field names more concise
+        boolean keepOriginalAlias = countDbSources(flattenedSourceDefinition) > 1;
```

Comment:
Personal opinion, but I could probably be swayed pretty easily: don't bother with this. I'd prefer less concise field names if it means simpler code and more consistent behavior.